### PR TITLE
Stdune Map/Set improvements

### PR DESCRIPTION
**Pull request by rgrinberg**
- [ ] re-implement sets in terms of maps. A set is now `type t = unit Map.t`. However, it could also be `type t = M : 'a Map.t -> t`. The advantage of this type would be that we could convert maps to sets for free.

- [x] add `OrderedType.t -> Dyn.t` as a requirement to the map functor. This will give us good error messages for find_exn for free and get rid of a bunch of pretty printing boilerplate we have everywhere.

- [x] Create Map/Set with a single functor application. These will now always need to agree on the comparison function.

- [ ] Implement a generic set parser for sexps. The idea here is that many fields in the dune language accept list valued arguments that cannot contain duplicates. These should really be parsed as sets. We'll parse them into `Loc.t Map.t` and then convert them to sets accordingly.

### CODE REVIEW

File: src/stdune/set.ml [hunk 1]
```diff
@@ -1,49 +1,71 @@
...
+
+  let to_list = M.keys
+
+  let is_empty = M.is_empty
+  let cardinal = M.cardinal
+  let equal = M.equal ~equal:(fun () () -> true)
+  let empty = M.empty
+  let singleton k = M.singleton k ()
+
+  let diff = M.merge ~f:(fun _ x y ->
```
Comment (aalekseyev) [inline src/stdune/set.ml]: These implementations of `diff` and `inter` are pretty inefficient. They're linear-time in the size of the larger map (in the worst and in the best case), whereas the old functions are linear-time in the size of the smaller map (or perhaps `O(m*log(n/m))`?) in the worst case and faster than that in the best case (when there's some clustering of keys).

I can't see enough public things exposed in `Map` to implement equally efficient versions though.

Not sure what we should do about this.

File: src/stdune/ordered.ml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module Make (Key : Map.Key) = struct
```
Comment (aalekseyev) [inline src/stdune/ordered.ml]: In base this is called `Comparable.Make`. I think we should try to use an existing name if we can.

File: src/stdune/set.ml (hunk 1 above)
Comment (rgrinberg) [reply src/stdune/set.ml]: Hmm. I believe the original idea was from @bobot. Let's see if has a solution for this

**Comment by aalekseyev** [comment]
By the way, one disadvantage of `type t = M : 'a Map.t -> t` is that it doesn't let `'a` values to be garbage-collected. (well, although you could still have a function `clean_up : Set.t -> Set.t` that throws them away if you really want to, so maybe it's strictly more powerful)

File: src/stdune/set.ml (hunk 1 above)
Comment (rgrinberg) [reply src/stdune/set.ml]: Another options is to vendor the map implementation from base or the stdlib. That shouldn't be so huge on its own, right?

Comment (aalekseyev) [reply src/stdune/set.ml]: I guess that can work, yeah. If we're doing a major change like that, I wish we used the more flexible types from Base which let you write functions generic over all maps. 

That said, base doesn't expose the efficient hedge-union operations either. We'd have to improve that.

Comment (rgrinberg) [reply src/stdune/set.ml]: > I guess that can work, yeah. If we're doing a major change like that, I wish we used the more flexible types from Base which let you write functions generic over all maps.

We should do that too. I believe we can do that separately however. We can introduce a polymorphic module for maps (like you did for hashtable) and we can slowly start porting things. Porting everything at once seems too hard.

Comment (aalekseyev) [reply src/stdune/set.ml]: > Porting everything at once seems too hard.

Oh, I thought Base.Map.Make existed which would give us a smooth way to migrate. but I guess I was thinking of `core_kernel`.

**Comment by rgrinberg** [comment]
> By the way, one disadvantage of type t = M : 'a Map.t -> t is that it doesn't let 'a values to be garbage-collected. (well, although you could still have a function clean_up : Set.t -> Set.t that throws them away if you really want to, so maybe it's strictly more powerful)

I agree that seems error prone. Another way would be to have a function like:

```ocaml
val as_set : map -> f:(set -> 'a) -> 'a
```

This way, the caller has to least to think about the lifetime of this set. Probably the simplest representation is good enough for now.

Anyway, none of this is essential to the PR. So I will go on and address your other points. We can change the representation of sets later.

**Comment by rgrinberg** [comment]
@aalekseyev I removed the map based implementation and adapted the names from base. I believe this should be ready now. The last bullet point I'll save for another PR and porting all the other functor applications to `Comparable.Make` can be done incrementally.

**Comment by diml** [comment]
Before we import the modules from Base, I remember discussing a more composable API for maps and sets in Base with someone at Jane Street. We didn't push the idea much as it would have been a difficult change to go through in Base, however given that Dune is closed code base it's a good place to try such ideas. I'll look for it in my emails.

**Comment by yminsky** [comment]
I'm not sure what you mean, but we did end up creating a more composable
set of Map API's for Base.  So now Maps and Sets no longer need to be
created via functor for each type.  Is that what you mean?

**Comment by diml** [comment]
It's regarding comparators. The idea is that if you want to create a set of `string option` it's a bit awkward to do with the first class module API. I dug up in my emails and found the discussion from two years ago with @aalekseyev and mgadd. We actually merged something in core_kernel, and for instance you can do this:

```ocaml
$ open Core_kernel;;
$ Option.comparator String.comparator;;
- : (string option, String.comparator_witness Option.comparator_witness)
    Comparator.t
= {Core_kernel__.Comparator.compare = <fun>; sexp_of_t = <fun>}
```

However, I can't figure out how to use the resulting value to create a map.

Anyway, my reasoning is that we could go for this API in Dune given that it looks more flexible and it is easier to do such changes in Dune given that our internal libraries have no user. Then if it works well for Dune and once it has properly matured, we could consider changing the Base API.

**Comment by aalekseyev** [comment]
Found it:

```
Base.Map.Using_comparator.of_alist ~comparator:(Option.comparator String.comparator)
```

**Pull request merged**